src/main/java/net/fabricmc/accesswidener/AccessWidenerRemapper.java

```diff
@@ -16,66 +16,104 @@
 
 package net.fabricmc.accesswidener;
 
-import java.util.Map;
-
 import org.objectweb.asm.commons.Remapper;
 
-public final class AccessWidenerRemapper {
-	private final AccessWidener input;
-	private final String to;
-	private final Remapper remapper;
+/**
+ * Decorates a {@link AccessWidenerVisitor} with a {@link Remapper}
+ * to remap names passing through the visitor if they come from a different namespace.
+ */
+public final class AccessWidenerRemapper implements AccessWidenerVisitor {
+	private final Provider remapperProvider;
+	private final String targetNamespace;
+	private final AccessWidenerVisitor delegate;
+	// is null while no header was read or when no remapping should occur
+	private Remapper remapper;
 
 	/**
-	 * @param input The access widener to remap. It will not be modified.
-	 * @param to    The namespace that the access widener will be remapped to by the given remapper.
+	 * @param delegate        The visitor to forward the remapped information to.
+	 * @param remapper        Will be used to remap names found in the access widener.
+	 * @param sourceNamespace The expected namespace of the access widener being remapped. Remapping fails if the
+	 *                        actual namespace is different.
+	 * @param targetNamespace The namespace that the access widener will be remapped to.
 	 */
-	public AccessWidenerRemapper(AccessWidener input, Remapper remapper, String to) {
-		this.input = input;
-		this.to = to;
-		this.remapper = remapper;
+	public AccessWidenerRemapper(
+			AccessWidenerVisitor delegate,
+			Remapper remapper,
+			String sourceNamespace,
+			String targetNamespace
+	) {
+		this.delegate = delegate;
+		this.targetNamespace = targetNamespace;
+		this.remapperProvider = (from, to) -> {
+			if (!sourceNamespace.equals(from)) {
+				throw new IllegalArgumentException("Cannot remap access widener from namespace '" + from + "'."
+						+ " Expected " + sourceNamespace);
+			}
+
+			return remapper;
+		};
 	}
 
 	/**
-	 * @return Either the original access widener if no remapping is necessary, or a new access widener that contains
-	 * remapped names.
+	 * @param delegate         The visitor to forward the remapped information to.
+	 * @param remapperProvider Will be used to acquire a remapper based on the namespace found in the access widener.
+	 *                         Can return null to indicate no remapping should occur for this namespace.
+	 * @param targetNamespace  The namespace that the access widener will be remapped to.
 	 */
-	public AccessWidener remap() {
-		// Dont remap if we dont need to
-		if (input.namespace.equals(to)) {
-			return input;
+	public AccessWidenerRemapper(
+			AccessWidenerVisitor delegate,
+			Provider remapperProvider,
```

Comment:
I dont see the need of the provider? Just seems to make it neednlessly complex for something we arent going to use?